File: backend/src/main/java/com/gxs/bank/service/PromotionService.java
package com.gxs.bank.service;

import com.gxs.bank.exception.ResourceNotFoundException;
import com.gxs.bank.model.Promotion;
import com.gxs.bank.repository.PromotionRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class PromotionService {

    private final PromotionRepository promotionRepository;

    public List<Promotion> getActivePromotions() {
        return promotionRepository.findByIsActiveTrueOrderByValidFromDesc();
    }

    public Promotion getPromotion(UUID id) {
        return promotionRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Promotion not found"));
    }
}

File: backend-python/com/gxs/bank/service/PromotionService.py
from com.gxs.bank.exception.ResourceNotFoundException import ResourceNotFoundException
from com.gxs.bank.repository.PromotionRepository import PromotionRepository


class PromotionService:
    def __init__(self, db):
        self.db = db
        self.promotionRepository = PromotionRepository(db)

    def getActivePromotions(self):
        return self.promotionRepository.findByIsActiveTrueOrderByValidFromDesc()

    def getPromotion(self, promotionId: str):
        promotion = self.promotionRepository.findById(promotionId)
        if promotion is None:
            raise ResourceNotFoundException("Promotion not found")
        return promotion
